### DISCUSSION TITLE
How to update current tempo/time signature in real time

### DISCUSSION THREAD

**Discussion by LeoMassaFM** [melanchall/drywetmidi#270] [Q&A] [answered]
Continuing from issue #201:

I got all the tempo changes of my MIDI file with `tempoChanges = tempoMap.GetTempoChanges();` and printed them all, along with their respective BPMs with the `BeatsPerMinute` property, with a foreach loop, which gives me the tempos and at what times they occur in the MIDI. I also created a public double `currentBPM` that holds the current BPM acquired from BeatsPerMinute.
For example, my MIDI has a tempo change of 140 BPM at 0 seconds (the very start of the file), a second one of 124 BPM that occurs 8 measures (32 quarter notes) after the first one, and a third one of 152 BPM that occurs 4 measures (16 beats) after the second one. 
How do I make `currentBPM`, which, in this case, starts at 140, change its value in the inspector during play mode to 124 when 32 beats pass after the first tempo change, and to 152 when 16 beats pass after the second tempo change?
Do I need to use the Playback function to achieve that? If not, how do I accomplish the task?
Thanks, and sorry for the wall of text. And apologies for the dumb question, I'm still learning how to use this plugin.

**Comment by melanchall** [comment, marked as answer]
Hi,

`Playback` is the best choice in my opinion. Please read the library documentation. You need to subscribe to the [`EventPlayed`](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Multimedia.Playback.html#Melanchall_DryWetMidi_Multimedia_Playback_EventPlayed) event and check if it's a `SetTempoEvent` one:

```csharp
// ...

_playback.EventPlayed += OnEventPlayed;

// ...

private static void OnEventPlayed(object? sender, MidiEventPlayedEventArgs e)
{
    if (e.Event is SetTempoEvent setTempoEvent)
    {
        var newBpm = new Tempo(setTempoEvent.MicrosecondsPerQuarterNote).BeatsPerMinute;
    }
}
```

But if you want to solve the task manually without `Playback`, I can't help you unfortunately. The area of responsibility of the DryWetMIDI ends when you get `tempoChanges`. What you're going to do with all those tempos is up to you. It's your part how to manage timings then.

**Reply by LeoMassaFM** [reply]
I'll try that. If something noteworthy turns up, I'll let you know. Thanks.

**Reply by LeoMassaFM** [reply]
If both a time signature and tempo change occur simultaneously, do I need a second `MidiEventPlayedEventArgs` variable or is it possible to just use a single one?
In case a second one isn't needed, in order to track time signature changes just like tempo changes, would something like this be correct?
```c#

    if (e.Event is SetTempoEvent setTempoEvent)
    {
        var newBpm = new Tempo(setTempoEvent.MicrosecondsPerQuarterNote).BeatsPerMinute;
        print(newBpm);       
    }


    if (e.Event is TimeSignatureEvent timeSignatureEvent)
    {
        var newTimeSig = new TimeSignature(timeSignatureEvent.Numerator, timeSignatureEvent.Denominator);
        print(newTimeSig.Numerator + "/" + newTimeSig.Denominator);
    }
```

**Reply by melanchall** [reply]
Yes, that's correct.

**Reply by LeoMassaFM** [reply]
This works all fine and dandy, but it doesn't stop when exiting play mode. I have `_playback.Loop` set to `true`. My project is, for the time being, a DDR clone, meaning the MIDI is used to spawn notes. I want the spawning notes and `OnEventPlayed` to execute indefinetely while in play mode and stop executing both functions if play mode is exited.

**Reply by melanchall** [reply]
@LeoMassaFM So just call `_playback.Stop()` inside the method which called by Unity on exiting play mode. I'm not an expert in Unity. I know there is the `Start` method there where you can put the code to execute on entering play mode. I suppose there is the method for exiting it.
